bugzoo/mgr/coverage/extractor.py

```diff
@@ -9,265 +9,96 @@
-
-        try:
-            if instrument:
-                logger.debug("instrumenting container")
-                self.prepare()
-            else:
-                logger.debug("not instrumenting container")
-        except Exception:
-            msg = "failed to instrument container."
-            raise exceptions.FailedToComputeCoverage(msg)
-
-        cov = {}
-        for test in _tests:
-            logger.debug("Generating coverage for test %s in container %s",
-                         test.name, container.uid)
-            outcome = self.__installation.containers.execute(container, test)
-            filelines = self.extract()
-            test_coverage = TestCoverage(test.name, outcome, filelines)
-            logger.debug("Generated coverage for test %s in container %s",
-                         test.name, container.uid)
-            cov[test.name] = test_coverage
-
-        self.cleanup()
-
-        coverage = TestSuiteCoverage(cov)
-        logger.debug("Computed coverage for container: %s", container.uid)
-        return coverage
-
-    def cleanup(self) -> None:
-        """
-        Strips instrumentation from the source code inside the container,
-        and reconfigures its program to no longer use coverage options.
-        """
-        logger.debug("cleanup method for gcov extractor does nothing.")
+def register_as_default(language: Language):
+    """
+    Registers a coverage extractor class as the default coverage extractor
+    for a given language.
+    """
+    def decorator(cls: Type['CoverageExtractor']):
+        cls.register_as_default(language)
+        return cls
+    return decorator
 
-    def extract(self) -> None:
-        """
-        Uses gcovr to extract coverage information for all of the C/C++ source
-        code files within the project. Destroys '.gcda' files upon computing
-        coverage.
-        """
-        logger_c = logger.getChild(container.id)  # type: logging.Logger
-        container = self.__container
-        mgr_ctr = self.__installation.containers
-        mgr_bug = self.__installation.bugs
-        logger_c.debug("Extracting coverage information")
 
-        bug = mgr_bug[container.bug]
-        dir_source = bug.source_dir
-        t_start = timer()
-        logger_c.debug("Running gcovr.")
-        fn_temp_ctr = mgr_ctr.mktemp(container)
-        cmd = 'gcovr -o "{}" -x -d -r .'.format(fn_temp_ctr)
-        response = mgr_ctr.command(container,
-                                   cmd,
-                                   context=dir_source,
-                                   verbose=True)
-        logger_c.debug("Finished running gcovr (took %.2f seconds).", timer() - t_start)  # noqa: pycodestyle
-        assert response.code == 0, "failed to run gcovr"
+class CoverageExtractor(object):
+    """
+    Used to compute coverage information for a given container according to a
+    provided set of instructions.
+    """
+    @classmethod
+    def register(cls, name: str) -> None:
```

Comment:
This function and the next one look very similar to the ones in `CoverageInstruction`. It's a bit confusing what an extractor is and what an instruction is and who is registering the name.

Reply:
Also after reading `GCovExtractor` I notice that there are function required by the API to be implemented by `CoverageExtractor`'s subclasses. Those should be part of `CoverageExtractor`.

Reply:
> Also after reading `GCovExtractor` I notice that there are function required by the API to be implemented by `CoverageExtractor`'s subclasses. Those should be part of `CoverageExtractor`.

I'm not quite sure what you mean. Could you elaborate?

Reply:
> This function and the next one look very similar to the ones in `CoverageInstruction`. It's a bit confusing what an extractor is and what an instruction is and who is registering the name.

We could rename `register` and `register_as_default` to `_register` and `_register_as_default`? These aren't supposed to be called by the user. Instead, the user should call the `@register` and `@register_as_default` decorators.

As for why this is necessary, we need a way to map `type` names in the manifest file to `CoverageInstruction` subclasses, which we then in turn need to map to their respective `CoverageExtractor` class. The implementation details are hidden from the user by requiring that they write a `CoverageExtractor` (with an internal `Instructions` class) and decorate that `CoverageExtractor` only once.

Reply:
> > Also after reading `GCovExtractor` I notice that there are function required by the API to be implemented by `CoverageExtractor`'s subclasses. Those should be part of `CoverageExtractor`.
> 
> I'm not quite sure what you mean. Could you elaborate?

I know what you mean. I've added abstract methods to `CoverageExtractor` and documented the API.